[CASE test_every_model_has_objects_queryset_available]
from django.db import models

class User(models.Model):
    pass

reveal_type(User.objects)  # E: Revealed type is 'django.db.models.query.QuerySet[main.User]'
reveal_type(User.objects.get())  # E: Revealed type is 'main.User*'