[case testForeignKeyWithClass]
from django.db import models

class Publisher(models.Model):
    pass

class Book(models.Model):
    publisher = models.ForeignKey(to=Publisher, on_delete=models.CASCADE)

book = Book()
reveal_type(book.publisher)  # E: Revealed type is 'main.Publisher*'
[out]


[case testForeignKeyRelatedName]
from django.db import models

class Publisher(models.Model):
    pass

class Book(models.Model):
    publisher = models.ForeignKey(to=Publisher, on_delete=models.CASCADE,
                                    related_name='books')

publisher = Publisher()
reveal_type(publisher.books)  # E: Revealed type is 'django.db.models.query.QuerySet[main.Book]'
[out]
