[case testBasicModelFields]
from django.db import models


class User(models.Model):
    id = models.AutoField(primary_key=True)
    small_int = models.SmallIntegerField()
    name = models.CharField(max_length=255)
    slug = models.SlugField(max_length=255)
    text = models.TextField()

user = User()
reveal_type(user.id)  # E: Revealed type is 'builtins.int'
reveal_type(user.small_int)  # E: Revealed type is 'builtins.int'
reveal_type(user.name)  # E: Revealed type is 'builtins.str'
reveal_type(user.slug)  # E: Revealed type is 'builtins.str'
reveal_type(user.text)  # E: Revealed type is 'builtins.str'
[out]
